blog/content/post/2023-06-07-twpm-spi-fix.md

```diff
@@ -0,0 +1,855 @@
+
+	// One-time SPI configuration
+	// Clear SPI_RXFIFO_THRESHOLD to trigger DMA on each byte available.
+	CLEAR_BIT(hspi->Instance->CR2, SPI_RXFIFO_THRESHOLD);
+
+	// Start the transfer
+	SET_BIT(hspi->Instance->CR2, SPI_CR2_RXDMAEN);
+	HAL_DMA_Start_IT(hspi->hdmarx, (uint32_t)&hspi->Instance->DR, (uint32_t)rx_buf, 4);
+
+	HAL_DMA_Start_IT(hspi->hdmatx, (uint32_t)tx_buf, (uint32_t)&hspi->Instance->DR, 4);
+	SET_BIT(hspi->Instance->CR2, SPI_CR2_TXDMAEN);
+
+	__HAL_SPI_ENABLE(hspi);
+}
+```
+
+The code size is reduced to almost a minimum - still, some optimizations could
+be done in `HAL_DMA_Start_IT`. Currently, we transmit only 4 bytes of static
+data to test whether MCU can handle this before going further.
+
+I'm using a bit different initialization sequence than HAL: HAL enables
+`RXDMAEN` after programming the channel and `TXDMAEN` after enabling SPI. Our
+code follows the sequence described in the STM32 Programming Manual (**rm0351**).
+
+
+![](/img/rm0351_spi_dma_enable_procedure.png)
+
+For testing purposes, I'm using Raspberry PI 3B as SPI host. Configuration is
+pretty straightforward, you can enable `spidev` by uncommenting
+`dtoverlay=spi0-1cs` in `/boot/config.txt` and rebooting. For communicating with
+`spidev` I'm using a custom Python script:
+
+```python
+from spidev import SpiDev
+
+class Spi:
+    def __init__(self):
+        self.device = SpiDev()
+        self.device.open(0, 0)
+        self.device.bits_per_word = 8
+        self.device.mode = 0b00
+        self.device.max_speed_hz = 24000000
+        self.freq = 24000000
+
+    def get_frequency(self):
+        return self.freq
+
+    def set_frequency(self, freq):
+        self.freq = freq
+
+    def xfer(self, data: bytes) -> bytes:
+        return self.device.xfer(data, self.freq)
+
+
+def test(func):
+    def wrapper():
+        freq = spi.freq
+        iteration = 0
+
+        try:
+            for i in range(10):
+                iteration = i
+                func()
+            print(f'OK: {func.__name__} @ {freq} Hz')
+        except AssertionError:
+            print(f'FAIL: {func.__name__} @ {freq} Hz (iteration {iteration})')
+    return wrapper
+
+
+@test
+def test_read_constant_4byte():
+    expected_data = [0xaa, 0xe0, 0xbb, 0xa5]
+    result = spi.xfer([0xff] * len(expected_data))
+    print('result = {:x} {:x} {:x} {:x}'.format(result[0], result[1], result[2], result[3]))
+    assert result == expected_data
+
+
+def main():
+    global spi
+
+    try:
+        spi = Spi()
+        spi.set_frequency(24000000)
+        test_read_constant_4byte()
+    finally:
+        spi.device.close()
+
+if __name__ == '__main__':
+    main()
+```
+
+After running the test code, I saw the transmitted data was correct through the
+logic analyzer, but Raspberry PI didn't receive the right data. This was a
+problem with the connection between Raspberry PI and Nucleo. I could achieve
+stable transmission at frequencies up to 18 MHz. After changing cable
```

Comment:
Changing how? Shorter cables? How short?

Reply:
I extended the description